### DISCUSSION TITLE
Getting a Starlette's UploadFile object instead of a FastApi's UploadFile in a POST endpoint

### DISCUSSION THREAD

**Discussion by Bktero** [fastapi/fastapi#12385] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import fastapi
import starlette.datastructures
from typing import Dict

app = fastapi.FastAPI()


@app.post("/upload/")
async def upload_file(file: fastapi.UploadFile) -> Dict[str, object]:
    response = {
        "Type": str(type(file)),
        "fastapi.UploadFile": isinstance(file, fastapi.UploadFile),
        "starlette.datastructures.UploadFile": isinstance(file, starlette.datastructures.UploadFile)
    }

    try:
        file.validate(str)
        response["AttributeError"] = "No error"
    except AttributeError as e:
        response["AttributeError"] = str(e)

    return response
```

### Description

Save the code to `api.py`

Check it with `mypy`:

```shell
$ mypy api.py
Success: no issues found in 1 source file
```

Run this code with `uvicorn api:app --reload --port 49000`

From another terminal, make a request:

```shell
$ echo "hello" > data.txt 
$ curl -X POST -L -F "file=@data.txt"  http://127.0.0.1:49000/upload
{"Type":"<class 'starlette.datastructures.UploadFile'>","fastapi.UploadFile":false,"starlette.datastructures.UploadFile":true,"AttributeError":"'UploadFile' object has no attribute 'validate'"}% 
```

The `file` parameter isn't an instance of `fastapi.UploadFile`. This is not what the prototype of the function says. I don't know whether this is a real issue, but it may be confusing.

### Operating System

Linux

### Operating System Details

I am using Debian Bookworm.

```shell
uname -a                             
Linux debian 6.10.6+bpo-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.10.6-1~bpo12+1 (2024-08-26) x86_64 GNU/Linux
```

### FastAPI Version

0.115.0

### Pydantic Version

2.9.2

### Python Version

Python 3.11.2

**Comment by JavierSanchezCastro** [comment]
Check this: https://github.com/fastapi/fastapi/discussions/12296

**Comment by Kfir-G** [comment]
Hi @Bktero, 

In FastAPI, the `UploadFile` class is actually a subclass of Starlette's `UploadFile`, which can cause confusion when performing type checks, as it may appear as a Starlette type, making it important to understand FastAPI’s reliance on Starlette for file uploads.

You can check further more [here](https://dev.to/kfir-g/understanding-fastapis-uploadfile-the-starlette-connection-4na2).

**Comment by Bktero** [comment]
I know that fastapi's `UploadFile` is a subclass of starlette's class with the same name. However, I don't see how this justify that the object being actually passed to the callback is of the type of the base class, not of the derived class.

Let's take this (overly?) simple example:
 
```python
class Base:
    pass


class Derived(Base):
    pass


def callback(o):
    print("Type =", type(o))
    print("Base =", isinstance(o, Base))
    print("Derived =", isinstance(o, Derived))


b = Base()
d = Derived()

callback(b)
callback(d)
```

This prints:

```
Type = <class '__main__.Base'>
Base = True
Derived = False
Type = <class '__main__.Derived'>
Base = True
Derived = True
```

The same should apply to UploadFile classes.

If I change the code to add a type hint : `def callback(o: Derived)`
Then mypy will complain: `error: Argument 1 to "callback" has incompatible type "Base"; expected "Derived"  [arg-type]`

**Comment by Bktero** [comment]
OK. I've dug a little into the source code.

Starlette creates the object, of course using it's own class `starlette.datastructures.UploadFile`. Then, at some some point, `fastapi.datastructures.UploadFile._validate` is called and casts this `starlette.datastructures.UploadFile` object to `fastapi.datastructures.UploadFile`.

So, and the end of the day, we have something like this, right?

```python
from typing import cast

class Base:
    pass


class Derived(Base):
    pass

def intermediate(o):
    c = cast(Derived, o)
    callback(c)

def callback(o: Derived):
    print("Type =", type(o))
    print("Base =", isinstance(o, Base))
    print("Derived =", isinstance(o, Derived))


b = Base()
intermediate(b)
```

mypy is happy, thanks to the cast.

From an OOP perspective, this cast is illegal. `Derived` is a `Base` to but `Base` it's not a `Derived`.

**Reply by sinisaos** [reply]
@Bktero I find this interesting and I was surprised by the behavior of FastAPI `UploadFile`. I also know that `UploadFile` is a subclass of Starlette `UploadFile`, but just like you, I thought that `type(UploadFile)` would be `<class 'fastapi.datastructures.UploadFile'>` and not `<class 'starlette.datastructures.UploadFile'>`. If that makes any sensce we can make workaround by instantiating `UploadFile` in route. After that everything is fine and we can use classmethod `validate`. Something like this
```python
import uvicorn
from fastapi import FastAPI, UploadFile
import starlette.datastructures
from typing import Dict

app = FastAPI()


@app.post("/upload/")
async def upload_file(file: UploadFile) -> Dict[str, object]:
    # instantiating again
    instance_file = UploadFile(file=file)
    response = {
        "Type": str(type(instance_file)),
        "fastapi.UploadFile": isinstance(instance_file, UploadFile),
        "starlette.datastructures.UploadFile": isinstance(
            instance_file, starlette.datastructures.UploadFile
        ),
    }
    try:
        instance_file.validate(file)
        response["ValueError"] = "No error"
    except ValueError as e:
        response["ValueError"] = str(e)

    return response


if __name__ == "__main__":
    uvicorn.run(app)
```
Result is

```bash
{
  "Type": "<class 'fastapi.datastructures.UploadFile'>",
  "fastapi.UploadFile": true,
  "starlette.datastructures.UploadFile": true,
  "ValueError": "No error"
}
```
